### ISSUE TITLE
Memory Leak with AmazonDynamoDBClientBuilder during Tomcat undeploy/deploy

### ISSUE THREAD

**Issue by wilkowski** [aws/aws-sdk-java#1994]
I've been trying to track down a memory leak on our tomcat system when deploying/undeploying servlets and it seems to be caused by building the AmazonDynamoDB client. I built a small test servlet and have been seeing a memory leak with every undeployment of it. On my http://localhost:8080/manager/html page, the 'Find leaks' button says that there is a memory leak with the servlet and I also see a bunch of com.amazonaws classes sticking around in VisualVM after a deploy+undeploy. I've uploaded the code I used for my test servlet [here](https://github.com/wilkowski/LeakTestConsole). 

#918 seems very similar so I left a comment there as well.

**Comment by zoewangg** [comment]
Hi @wilkowski , can you share the logs or stacktrace showing the memory leak?

In your [test code](https://github.com/wilkowski/LeakTestConsole/blob/3ef58d97dd1e3086fb0f11d775b99eadd18b853b/src/com/leaktestconsole/ServletInit.java#L27 ), 

```
	@Override
	public void contextDestroyed(ServletContextEvent arg0) {
		// It definitely did not work with client.shutdown() here
}
```

What do you mean by `It definitely did not work with client.shutdown() here`? What would happen if you invoke client.shutdown there?

**Comment by wilkowski** [comment]
I tried making AmazonDynamoDB client a class variable and calling client.shutdown() from contextDestroyed but it had no noticeable effect on the memory leak. I used VisualVM to take a snapshot of my tomcat process after deploying and undeploying my test console. I have a screenshot [here](https://imgur.com/a/Zo5eisU) of the com.amazonaws classes that were still loaded after undeploying. My primary way of detecting the leak has been the 'Find leaks' diagnostic tool on my tomcat web application manager page. I can't provide a full heap dump since it could contain my ec2 credentials.
